crates/sui-core/src/traffic_controller.rs

```diff
@@ -0,0 +1,169 @@
+                proxy_ips: Arc::new(RwLock::new(HashMap::new())),
+            },
+        };
+        let blocklists = ret.blocklists.clone();
+        spawn_monitored_task!(run_tally_loop(rx, policy_config, blocklists));
+        ret
+    }
+
+    pub fn tally(&self, tally: TrafficTally) {
+        // Use try_send rather than send mainly to avoid creating backpressure
+        // on the caller if the channel is full, which may slow down the critical
+        // path. Dropping the tally on the floor should be ok, as in this case
+        // we are effectively sampling traffic, which we would need to do anyway
+        // if we are overloaded
+        match self.tally_channel.try_send(tally) {
+            Err(TrySendError::Full(_)) => {
+                warn!("TrafficController tally channel full, dropping tally");
+                // TODO: metric
+            }
+            Err(TrySendError::Closed(_)) => {
+                panic!("TrafficController tally channel closed unexpectedly");
+            }
+            Ok(_) => {}
+        }
+    }
+
+    pub async fn check(
+        &self,
+        connection_ip: Option<SocketAddr>,
+        proxy_ip: Option<SocketAddr>,
+    ) -> bool {
+        match (connection_ip, proxy_ip) {
+            (Some(connection_ip), _) => self.check_and_clear_blocklist(connection_ip, true).await,
+            (_, Some(proxy_ip)) => self.check_and_clear_blocklist(proxy_ip, false).await,
+            _ => true,
+        }
+    }
+
+    async fn check_and_clear_blocklist(&self, ip: SocketAddr, connection_ips: bool) -> bool {
+        let blocklist = if connection_ips {
+            self.blocklists.connection_ips.clone()
+        } else {
+            self.blocklists.proxy_ips.clone()
+        };
+
+        let now = SystemTime::now();
+        let expiration = blocklist.read().get(&ip).copied();
+        match expiration {
+            Some(expiration) if now >= expiration => {
+                blocklist.write().remove(&ip);
+                true
+            }
+            None => true,
+            _ => false,
+        }
+    }
+}
+
+async fn run_tally_loop(
+    mut receiver: mpsc::Receiver<TrafficTally>,
+    policy_config: PolicyConfig,
+    blocklists: Blocklists,
+) {
+    let mut spam_policy = policy_config.clone().to_spam_policy();
+    let mut error_policy = policy_config.clone().to_error_policy();
+    let spam_blocklists = Arc::new(blocklists.clone());
+    let error_blocklists = Arc::new(blocklists);
+    loop {
+        tokio::select! {
+            received = receiver.recv() => match received {
+                Some(tally) => {
+                    handle_spam_tally(&mut spam_policy, &policy_config, tally.clone(), spam_blocklists.clone()).await;
+                    handle_error_tally(&mut error_policy, &policy_config, tally, error_blocklists.clone()).await;
+                }
+                None => {
+                    panic!("TrafficController tally channel closed unexpectedly");
```

Comment:
i don't think this panic is necessary - the panic on the sender side if the channel is closed should be sufficient. this one might just cause problems in tests when we drop the authority.